tests/unit_tests/service/test_authentication.py

```diff
@@ -0,0 +1,171 @@
+    return SessionManager.from_config(mock_server_config, mock_client_config)
+
+
+@pytest.fixture
+def mock_connected_client_config(mock_client_config: OAuthClientConfig):
+    assert isinstance(mock_client_config, CLIClientConfig)
+    with open(mock_client_config.token_file_path, "w") as token_file:
+        # base64 encoded token
+        token_file.write(
+            base64.b64encode(
+                b'{"access_token":"token","refresh_token":"refresh_token"}'
+            ).decode("utf-8")
+        )
+    return mock_client_config
+
+
+@pytest.fixture
+def mock_authenticator(mock_server_config, mock_client_config) -> Authenticator:
+    return Authenticator(mock_server_config, mock_client_config)
+
+
+@mock.patch("jwt.decode")
+@mock.patch("jwt.PyJWKClient.get_signing_key_from_jwt")
+def test_verify_token_valid(
+    mock_get_signing_key, mock_decode, mock_authenticator: Authenticator
+):
+    decode_retun_value = {"token": "valid_token", "name": "John Doe"}
+    mock_decode.return_value = decode_retun_value
+    valid_token = mock_authenticator.verify_token(decode_retun_value["token"])
+    assert valid_token
+
+
+@mock.patch("jwt.decode")
+@mock.patch("jwt.PyJWKClient.get_signing_key_from_jwt")
+def test_verify_token_invalid(
+    mock_get_signing_key, mock_decode, mock_authenticator: Authenticator
+):
+    mock_decode.side_effect = jwt.ExpiredSignatureError
+    token = "invalid_token"
+    with pytest.raises(PyJWTError):
+        mock_authenticator.verify_token(token)
+
+
+@mock.patch("jwt.decode")
+@mock.patch("jwt.PyJWKClient.get_signing_key_from_jwt")
+def test_user_info(
+    mock_get_signing_key,
+    mock_decode,
+    mock_authenticator: Authenticator,
+):
+    mock_decode.return_value = {
+        "name": "John Doe",
+        "fedid": "12345",
+    }
+    mock_authenticator.print_user_info("valid_token")
+
+
+def test_logout(
+    mock_session_manager: SessionManager, mock_connected_client_config: CLIClientConfig
+):
+    assert os.path.exists(mock_connected_client_config.token_file_path)  # type: ignore
+    mock_session_manager.logout()
+    assert not os.path.exists(mock_connected_client_config.token_file_path)  # type: ignore
+
+
+@mock.patch("requests.post")
+def test_refresh_auth_token(
+    mock_post,
+    mock_session_manager: SessionManager,
+    mock_connected_client_config: OAuthClientConfig,
+):
+    mock_post.return_value.status_code = HTTPStatus.OK
+    mock_post.return_value.json.return_value = {"access_token": "new_access_token"}
+    result = mock_session_manager.refresh_auth_token()
+    assert result == {"access_token": "new_access_token"}
+
+
+@mock.patch("requests.post")
+def test_get_device_code(
+    mock_post,
+    mock_session_manager: SessionManager,
+):
+    mock_post.return_value.status_code = HTTPStatus.OK
+    mock_post.return_value.json.return_value = {"device_code": "device_code"}
+    device_code = mock_session_manager.get_device_code()
+    assert device_code == "device_code"
```

Comment:
Can we get a more complete test where we mock the poll_for_updates, and do slightly more of the flow? This is basically doing

a = b
assert a == b